### PR TITLE
Add native projection-MPO assembly utilities

### LINKED ISSUE #57: Add native projection-MPO assembly utilities
Parent epic: #55

Planned PR branch/name: `pr-2-projection-mpo-core`

## Scope
- Add internal helpers for building diagonal projection MPOs from sparse nonzero tensor entries.
- Port CoTenN utility ideas into clean Julia: `itensor_from_nonzeros`, tensor-to-MPO assembly, identity tensors, and pass-through tensors.
- Remove all debug printing from the adapted design.
- Add internal helpers `projection_mpo(constraint, sites)` and `projection_mpos(constraints, sites)`.

## Acceptance Criteria
- [ ] Small manually constructed projection MPOs materialize to diagonal masks.
- [ ] Projection MPOs preserve feasible basis states and zero infeasible ones.
- [ ] No dependency on CoTenN Python or local ITensorMPS forks.
- [ ] Attribution comment references CoTenN's projection-MPO construction.

## Depends On
- #56 (Add binary constraint types and feasibility checks)

## Notes
Each issue is intended to be completed by one focused PR.
## Attribution
This work is inspired by the projection-MPO constraint framework in Sharma, Ritvik, Cheng Peng, Siddharth Dangwal, and Sara Achour. "CoTenN: Constrained Optimization with Tensor Networks." Proceedings of the ACM on Programming Languages 10, no. PLDI (2026): 629-655.

### PR DESCRIPTION
**Pull request by bernalde**
## Summary

- Adds internal sparse ITensor construction and tensor-to-MPO assembly helpers for diagonal projection masks.
- Adds `projection_mpo(constraint, sites)` and `projection_mpos(constraints, sites)` for binary constraints over Qudit sites.
- Adds focused tests that materialize small projection masks and compare feasible/infeasible basis-state behavior against `is_feasible`.

Closes #57.

## Tests

- `julia --project=/tmp/tensolver-issue57-testenv -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed.
- `julia --project=/tmp/tensolver-issue57-testenv -e 'using Test, TenSolver; include("test/constraints.jl"); include("test/projection_mpo.jl")'` passed.
- `julia --project=. -e 'using Pkg; Pkg.test()'` passed.

## Branch Hygiene

- Base branch: `pr-1-constraint-api` (temporary stacked base for PR #84 / issue #56; retarget to `main` after #84 merges)
- Source branch point: `origin/pr-1-constraint-api` at `f6a2f80a6c2cd46d97d21cd27fb0885f92bcf9c6`
- Stacked status: stacked on PR #84 by issue request; opened as draft
- Prerequisite PRs: #84

### CODE REVIEW

File: src/projection_mpo.jl [hunk 1]
```diff
@@ -0,0 +1,196 @@
+# Projection-MPO construction adapted from the CoTenN constraint projection
+# design, but implemented natively with ITensors/ITensorMPS.
+
+struct _SparseTensorEntry{T}
+  coordinates::Dict{Int,Int}
+  value::T
+end
+
+"""
```
Comment (bernalde) [inline src/projection_mpo.jl:29]: Blocking: this new docstring is not included in the manual, and Documenter fails `docs/make.jl` with `missing_docs` for this helper plus `projection_mpo` and `projection_mpos`. Since these are intended internal helpers, either convert these docstrings to ordinary comments or add canonical docs entries for them. This is currently failing the documentation CI build.

**Review by bernalde — COMMENTED** [review]
Blocking:
- The documentation build fails because this PR adds module docstrings for internal helpers without adding them to canonical docs. `julia --project=docs/ docs/make.jl` fails with Documenter `missing_docs` for `TenSolver.projection_mpo`, `TenSolver.itensor_from_nonzeros`, and `TenSolver.projection_mpos`; the GitHub documentation build fails for the same reason. Fix by either converting these internal-helper docstrings to regular comments, or adding a canonical `@docs`/manual entry if they are meant to be documented.

Nonblocking:
- None.

Questions:
- None.

Tests run and outcomes:
- Verified merge-base diff totals against PR metadata: 4 files, +270/-0.
- `julia --project=/tmp/tensolver-pr85-review-testenv -e 'using Pkg; Pkg.develop(path=pwd()); Pkg.instantiate(); using Test, TenSolver; include("test/projection_mpo.jl")'` passed.
- `julia --project=. -e 'using Pkg; Pkg.test()'` passed from `/tmp/tensolver-pr85-review`.
- `julia --project=docs/ docs/make.jl` failed with Documenter `missing_docs` for the three new docstrings.
- GitHub CI: all Julia test matrix jobs pass; documentation `build` fails.

Merge-readiness:
- I would not merge this until the blocking issues above are addressed and documentation CI is green. Because I authored this PR, GitHub only permits a COMMENT review from me; the formal change-request verdict must come from another maintainer.

File: src/projection_mpo.jl (hunk 1 above)
Comment (bernalde) [inline src/projection_mpo.jl:29]: Blocking: this docstring (and those on `projection_mpo` at line 166 and `projection_mpos` at line 184) are not referenced in `docs/src/api.md`. Documenter's default `checkdocs = :all` then aborts the build: `julia --project=docs docs/make.jl` fails with `[:missing_docs]` for these three symbols, which breaks the Documentation CI workflow that runs on every PR. `Pkg.test()` does not catch this (it only runs `doctest()`, not `checkdocs`). Fix: add the three to `docs/src/api.md` under "Internal Functions", or convert these to plain `#` comments.

**Review by bernalde — COMMENTED** [review]
Senior maintainer review of #85 (native projection-MPO assembly utilities).

Verdict mechanics: I am the PR author, so GitHub rejects APPROVE/REQUEST_CHANGES from me here (HTTP 422); submitting as COMMENT. There is a blocking issue below, so the substantive verdict is request-changes; the formal change request must come from another maintainer.

Reviewed the PR head (9fc3905) against merge-base f6a2f80 (the head of `pr-1-constraint-api`, this PR's stacked base). File/line totals match the reported 4 files / +270 / -0. Note the base is `pr-1-constraint-api`, not `main`; see the stacked-base note under Questions.

## Intent vs issue #57
The PR delivers the issue's scope: `itensor_from_nonzeros`, tensor-to-MPO assembly, and `projection_mpo`/`projection_mpos`, all internal (not exported), with no wiring into `minimize`/`maximize` (scope guard respected). I independently verified the core behavioral claim beyond the supplied tests: the materialized diagonal equals `is_feasible` for constraints that exclude the first physical site (`relation(2<=3)`, `not_equals([2,3])`), out-of-order sites (`sum([3,1])`), a 4-site register with a middle subset, and a single-site register. All correct. `Closes #57` is appropriate for the eventual merge into `main`.

## Blocking
1. The Documentation workflow (`.github/workflows/documentation.yml`, which runs on every `pull_request`) fails on this PR. It runs `Pkg.develop` + `julia --project=docs/ docs/make.jl`, and `makedocs` aborts with:

```
Error: 3 docstrings not included in the manual:
    TenSolver.projection_mpo
    TenSolver.itensor_from_nonzeros
    TenSolver.projection_mpos
ERROR: LoadError: `makedocs` encountered an error [:missing_docs] -- terminating build before rendering.
```

Documenter's default `checkdocs = :all` requires every docstring in the module to appear in a canonical `@docs`/`@autodocs` block. These three helpers gained docstrings but no `docs/src/api.md` entry. This is the same convention enforced for #84, just in the other direction. It matters because it breaks the docs CI gate (and gh-pages deploy on merge). Importantly, `Pkg.test()` does NOT catch it: the test suite only runs `doctest()`, not the `checkdocs` pass, which is why the full suite is green while the docs build is red. Fix: add the three symbols to `docs/src/api.md` under "Internal Functions" (alongside `tensorize`, `qmatrix_permutation`, `preprocess_qubo`), or convert the three docstrings to plain `#` comments. I reproduced both the failure and that adding api.md entries is the consistent fix.

## Nonblocking
1. Dead code. `_identity_tensor` (line 33), `_pass_through_tensor` (line 40), and `_path_count` (line 52, called only by the dead `_pass_through_tensor`) are never invoked anywhere in `src/` or `test/`. `_tensor_to_mpo` reimplements the identity/pass-through logic inline (the `states = (1, 2)` branch plus `_mpo_coordinate`), so there are two parallel implementations of the same concept, one used and tested, one unused and untested. Issue #57 lists these as ideas to port, but untested dead code in `src/` rots. Either route `_tensor_to_mpo`'s per-site construction through these helpers, or drop them.
2. Missing tests for the validation/error paths this PR introduces: `itensor_from_nonzeros` `DimensionMismatch` (coordinate length mismatch), `_tensor_to_mpo` `ArgumentError` on empty sites and on non-binary (`dim != 2`) sites, and `projection_mpo`'s out-of-register `BoundsError`. A few `@test_throws` would lock in this behavior.
3. Attribution completeness. Issue #57 supplies a full citation (Sharma et al., "CoTenN: Constrained Optimization with Tensor Networks", PLDI 2026). The source comment (line 1) only says "adapted from the CoTenN constraint projection design" without the reference. Adding the full citation in the comment matches the issue's stated acceptance criterion and the precedent.

## Questions
1. Bond-dimension scaling. The MPO link dimension is the number of feasible assignments over the constraint's sites (up to `2^k` for `k` constraint sites), and that dimension is carried on every bond of the whole register regardless of distance from the constraint. Correct for the small masks here, but is an uncompressed `2^k` bond acceptable for the downstream consumer (#58), or should this be bounded/compressed before it is wired into solves?
2. Stacked base. The PR targets `pr-1-constraint-api`, not `main`. Merging into that branch will not auto-close #57 (GitHub auto-closes only on default-branch merge), and the description plans to retarget to `main` after #84 merges, at which point `Closes #57` resolves correctly. Confirming that retarget is the intended path so the link behaves as expected.

## Tests run (clean checkout of PR head, Julia 1.10.11)
- Targeted `test/projection_mpo.jl`: 37/37 pass.
- Full `Pkg.test()`: passed. Projection MPOs 37/37, Constraints 51/51, QUBODrivers 171/171, Aqua 11/11 (no new ambiguities/unbound args from the added methods), Doctests 1/1.
- Independent correctness checks (edge cases listed above): all match `is_feasible`.
- `julia --project=docs docs/make.jl`: FAILS with `[:missing_docs]` (the blocking issue).

## Merge readiness
Not mergeable as-is: the docs CI gate is broken by the missing api.md entries. The implementation itself is correct and well-tested for behavior; once the docs build is fixed (and ideally the dead code and error-path tests addressed), it would be in good shape. I would not merge this until the blocking issue above is addressed. As I authored the PR, the formal approval or change request must come from another maintainer.

File: src/projection_mpo.jl [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    selector = map(i -> ind_tuple[i] => coordinate_tuple[i], eachindex(ind_tuple))
+    tensor[selector...] = tensor[selector...] + convert(T, value)
+  end
+
+  return tensor
+end
+
+itensor_from_nonzeros(inds, nonzeros) = itensor_from_nonzeros(Float64, inds, nonzeros)
+
+function _identity_tensor(::Type{T}, site) where {T}
```
Comment (bernalde) [inline src/projection_mpo.jl]: Nonblocking (dead code): `_identity_tensor` here, `_pass_through_tensor` (line 40), and `_path_count` (line 52, used only by `_pass_through_tensor`) are never called. `_tensor_to_mpo` reimplements the pass-through/identity logic inline instead. Untested dead code; either use these helpers in the per-site construction or remove them.

File: src/projection_mpo.jl [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+function _tensor_to_mpo(::Type{T}, entries, sites) where {T}
+  site_vec = collect(sites)
+  isempty(site_vec) && throw(ArgumentError("sites must not be empty"))
+  all(site -> ITensors.dim(site) == 2, site_vec) ||
+    throw(ArgumentError("projection MPO sites must be binary Qudit indices"))
+
+  entry_vec = collect(entries)
+  _validate_sparse_entries(entry_vec, site_vec)
+
+  path_count = max(length(entry_vec), 1)
```
Comment (bernalde) [inline src/projection_mpo.jl]: Question: `path_count` equals the number of feasible assignments (up to `2^k` for `k` constraint sites) and is used as the dimension of every link across the whole register. Correct for small masks, but is an uncompressed `2^k` bond acceptable for the #58 consumer, or should bond dimension be bounded/compressed before this feeds into solves?

File: test/projection_mpo.jl [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    diagonal[bits] = real(ITensors.inner(adjoint(psi), H, psi))
+  end
+
+  return diagonal
+end
+
+@testset "Projection MPOs" begin
+  sites = ITensors.siteinds("Qudit", 3; dim=2)
+
+  @testset "Sparse ITensor construction" begin
```
Comment (bernalde) [inline test/projection_mpo.jl:36]: Nonblocking (missing tests): the error paths added by this PR are untested -- `itensor_from_nonzeros` `DimensionMismatch` on a coordinate-length mismatch, `_tensor_to_mpo` `ArgumentError` on empty or non-binary (`dim != 2`) sites, and `projection_mpo`'s out-of-register `BoundsError`. A few `@test_throws` here would lock in that behavior.

File: src/projection_mpo.jl [hunk 5]
```diff
@@ -0,0 +1,196 @@
+# Projection-MPO construction adapted from the CoTenN constraint projection
```
Comment (bernalde) [inline src/projection_mpo.jl:1]: Nonblocking (attribution): issue #57's acceptance criteria provide the full citation (Sharma et al., "CoTenN: Constrained Optimization with Tensor Networks", PLDI 2026). Consider including it here rather than just "the CoTenN constraint projection design".

**Comment by bernalde** [comment]
Commits pushed:
- `e0f484a` `fix: address projection MPO review feedback (#57)`

Main changes made:
- Added the new projection-MPO docstrings to `docs/src/api.md` so Documenter `checkdocs` no longer fails.
- Routed unconstrained MPO sites through `_pass_through_tensor`, with `_identity_tensor` used by the no-link pass-through case, so those helpers are no longer dead code.
- Added validation tests for sparse coordinate length mismatch, empty MPO sites, non-binary site dimensions, and out-of-register projection constraints.
- Expanded the source attribution comment with the CoTenN citation from issue #57.

Tests run and results:
- `julia --project=/tmp/tensolver-pr85-comments-testenv -e 'using Pkg; Pkg.develop(path=pwd()); Pkg.instantiate(); using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 41/41.
- `julia --project=/tmp/tensolver-pr85-comments-docenv docs/make.jl` passed; the previous `missing_docs` failure is gone.
- `julia --project=/tmp/tensolver-pr85-comments-testenv -e 'using Pkg; Pkg.test("TenSolver")'` passed.

Comments intentionally not addressed:
- The bond-dimension scaling question did not get a code change in this PR. This PR keeps the exact uncompressed projection-MPO core requested by #57; bounding/compression should be decided when #58 wires these MPOs into solves and can measure the downstream cost.

Remaining risks or follow-up items:
- GitHub CI for commit `e0f484a` is still pending at the time of this comment.

File: src/projection_mpo.jl (hunk 1 above)
Comment (bernalde) [reply src/projection_mpo.jl:29]: Addressed in e0f484a: added `itensor_from_nonzeros`, `projection_mpo`, and `projection_mpos` to `docs/src/api.md`. `julia --project=/tmp/tensolver-pr85-comments-docenv docs/make.jl` now passes.

Comment (bernalde) [reply src/projection_mpo.jl:29]: Addressed in e0f484a: added the three symbols to the Internal Functions `@docs` block in `docs/src/api.md`. The full docs build now passes locally.

File: src/projection_mpo.jl (hunk 2 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Addressed in e0f484a: `_tensor_to_mpo` now delegates unconstrained sites to `_pass_through_tensor`, and `_pass_through_tensor` uses `_identity_tensor` for the no-link case.

File: src/projection_mpo.jl (hunk 3 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Not changed in e0f484a. This PR keeps the exact uncompressed core projection MPO for #57; bounding/compression should be decided in #58 when these MPOs are wired into solves and the downstream cost is measurable.

File: test/projection_mpo.jl (hunk 4 above)
Comment (bernalde) [reply test/projection_mpo.jl:36]: Addressed in e0f484a: added `Validation errors` tests for all four error paths. Targeted `test/projection_mpo.jl` now passes 41 assertions.

File: src/projection_mpo.jl (hunk 5 above)
Comment (bernalde) [reply src/projection_mpo.jl:1]: Addressed in e0f484a: expanded the source attribution comment with the Sharma/Peng/Dangwal/Achour CoTenN PLDI 2026 citation from issue #57.

File: src/projection_mpo.jl (hunk 3 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Leaving this open as a tracked design item rather than resolving it, since the code is unchanged. Concrete ways to address it, to be decided in #58 when projections meet solves and the cost is measurable:

- Post-assembly truncation: after `_tensor_to_mpo`, call `ITensorMPS.truncate!(H; cutoff=...)` (or `orthogonalize!` + a `maxdim`). A diagonal 0/1 projector is exactly representable with bond dimension equal to the number of distinct feasible *prefixes* of the assignment, which is usually far below `2^k`, so an exact-cutoff truncation should shrink the link with no loss.
- Build compressed directly: merge paths that share a prefix/suffix (a finite-automaton-style construction over the constraint sites) so the link never reaches `2^k` in the first place.
- Benchmark the resulting bond against the DMRG `maxdim` schedule in #58 before picking a default `cutoff`/`maxdim`.

If we want a cheap guard now, the smallest step is an exact `truncate!(H; cutoff=0)` pass inside `projection_mpo`, but I would rather make that call in #58 with real downstream numbers. Tracking there.

**Comment by bernalde** [comment]
Rebased PR #85 onto updated `pr-1-constraint-api` after PR #84 review fixes.

Commits now on the branch:
- `b75d303` feat: add projection MPO assembly utilities (#57)
- `35f0e35` fix: address projection MPO review feedback (#57)
- `34a2594` test: update projection constraints for constructor API

Additional rebase fix:
- Updated `test/projection_mpo.jl` to use direct constraint constructors.
- Replaced the strict `<` zero-mask example with a supported infeasible `<= -1` `SumConstraint`.

Verification:
- `julia --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'`: passed, Projection MPOs 41/41.
- `julia --project=. -e 'using Pkg; Pkg.test()'`: passed. Constraints 53/53, Projection MPOs 41/41, QUBODrivers 169/169, Aqua 11/11, Doctests 1/1. Real Models still has 3 pre-existing broken tests.

File: src/projection_mpo.jl [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  return projection_mpo(T, constraint, constraint_sites(constraint), sites)
+end
+
+function projection_mpo(::Type{T}, constraint::ExactlyOneConstraint, sites) where {T}
+  return projection_mpo(T, constraint, constraint_sites(constraint), sites)
+end
+
+function projection_mpo(::Type{T}, constraint::RelationConstraint, sites) where {T}
+  return projection_mpo(T, constraint, constraint_sites(constraint), sites)
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: These functions are exactly the same. We can just have the AbstractContraint version.

```suggestion
```

File: src/projection_mpo.jl [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  projection_mpo(Float64, constraint, sites)
+
+projection_mpo(constraint::NotEqualsConstraint, sites) =
+  projection_mpo(Float64, constraint, sites)
+
+projection_mpo(constraint::ExactlyOneConstraint, sites) =
+  projection_mpo(Float64, constraint, sites)
+
+projection_mpo(constraint::RelationConstraint, sites) =
+  projection_mpo(Float64, constraint, sites)
```
Comment (iagoleal) [inline src/projection_mpo.jl]: These functions are also exactly the same. Let's change for a single path on the supertype.

```suggestion
projection_mpo(constraint::AbstractConstraint, sites) =
  projection_mpo(Float64, constraint, sites)
```

File: src/projection_mpo.jl [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+"""
+    projection_mpos([T], constraints, sites)
+
+Build one projection MPO per constraint.
+"""
+function projection_mpos(::Type{T}, constraints::AbstractVector{<:AbstractConstraint}, sites) where {T}
+  site_vec = collect(sites)
+
+  return [projection_mpo(T, constraint, site_vec) for constraint in constraints]
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl:237]: Why are we doing this one separately from `projection_mpo` instead of relying on multiple dispatch? Is it for type stability? If it affects type stability, it is good to have them separately. I would also suggest having test sets explicitly checking for this stability using `[Test.@inferred](https://docs.julialang.org/en/v1/stdlib/Test/#Test.@inferred)`

File: src/projection_mpo.jl [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+  return entries
+end
+
+"""
+    projection_mpo([T], constraint, sites)
+
+Build a diagonal projection MPO that preserves basis states satisfying
+`constraint` and zeros infeasible states.
+"""
```
Comment (iagoleal) [inline src/projection_mpo.jl]: This docstring is too terse. It is hard to understand what the function does. Let's improve upon it for readability. What does a user / developer have to know about `projection_mpo`?

File: src/projection_mpo.jl [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+  return projection_entries(T, constraint, constraint_sites(constraint))
+end
+
+function projection_entries(::Type{T}, constraint::ExactlyOneConstraint) where {T}
+  return projection_entries(T, constraint, constraint_sites(constraint))
+end
+
+function projection_entries(::Type{T}, constraint::RelationConstraint) where {T}
+  return projection_entries(T, constraint, constraint_sites(constraint))
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: They are all calling the abstract version. We can just let dispatch handle it.

```suggestion
```

File: src/projection_mpo.jl [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+function projection_entries(::Type{T}, constraint::AbstractConstraint, constraint_sites) where {T}
```
Comment (iagoleal) [inline src/projection_mpo.jl]: This function should have a docstring or comment explaining what it does.

File: src/projection_mpo.jl [hunk 12]
```diff
@@ -0,0 +1,245 @@
+# Projection-MPO construction adapted from the CoTenN constraint projection
+# design in Sharma, Ritvik, Cheng Peng, Siddharth Dangwal, and Sara Achour,
+# "CoTenN: Constrained Optimization with Tensor Networks," PLDI 2026.
+
+struct SparseTensorEntry{T}
+  coordinates::Dict{Int,Int}
+  value::T
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl:27]: I do not understand what this type is for. What is relation between the dict of coordinates and the value it holds? Is it some kind of permutation? We need documentation for this type. Also, why is it necessary?

**Review by iagoleal — CHANGES_REQUESTED** [review]
This is an initial review. It is getting a good shape already! However, we will probably have to do some iterations on this one.

Some general comments:
- There are many repetitive code paths. We can make use of subtyping, multiple dispatch or just plain function abstraction to simplify large chunks of code.
- The overall documentation is lacking. This comes from very bleeding-edge research and the methods should be thoroughly documented. This means both docstrings as well as helpful comments explaining "why" we choose a certain construction;
- We define many "manual" tensor building methods. Let's check the ITensors.jl and ITensorMPS.jl documentation to see if they already provide helpers that we could use to simplify our code.

File: src/projection_mpo.jl [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+function projection_mpo(::Type{T}, constraint::ExactlyOneConstraint, sites) where {T}
+  return projection_mpo(T, constraint, constraint_sites(constraint), sites)
+end
+
+function projection_mpo(::Type{T}, constraint::RelationConstraint, sites) where {T}
+  return projection_mpo(T, constraint, constraint_sites(constraint), sites)
+end
+
```
Comment (iagoleal) [inline src/projection_mpo.jl:206]:
```suggestion
```

File: src/projection_mpo.jl [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+  return constraint.sites
+end
+
+function constraint_sites(constraint::ExactlyOneConstraint)
+  return constraint.sites
+end
+
+function constraint_sites(constraint::RelationConstraint)
+  return [constraint.left_site, constraint.right_site]
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: If we move this to `constraints.jl`, we can simplify the API here by making it only depend on the abstract type `AbstractConstraint`. This way, if we ever add new constraint types, there will be no need to edit the projection MPO logic.

**Comment by bernalde** [comment]
Commits pushed:
- `7732e5c` `fix: document projection MPO internals`

Main changes made:
- Documented `SparseTensorEntry` and added it to the internal API docs.
- Expanded `projection_mpo` and `projection_mpos` docstrings with binary-register semantics, site indexing, and the exact uncompressed path construction.
- Added comments explaining feasible-assignment enumeration and how missing sparse coordinates become pass-through MPO tensors.
- Added `@inferred` tests for typed and default `projection_mpo`/`projection_mpos` calls.

Tests run and results:
- `julia +release --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 45/45.
- `julia +release --project=docs -e 'using Pkg; Pkg.instantiate(); include("docs/make.jl")'` passed; local deploy auto-detection warning only.
- `julia +release --project=. -e 'using Pkg; Pkg.test()'` passed; Real Models still has 3 expected `broken` tests.
- GitHub checks for `7732e5c` passed: Documentation, `documenter/deploy`, and the full Run tests matrix.

Comments intentionally not addressed:
- Bond-dimension compression remains a #58 design item. This commit documents that the current projector construction is exact and uncompressed instead of changing the construction/compression policy here.

Remaining risks or approval gates:
- PR state is clean and CI is green, but `reviewDecision` is still `CHANGES_REQUESTED` until a maintainer updates the review state.

File: src/projection_mpo.jl (hunk 3 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Still intentionally left open for #58. `7732e5c` documents that the current construction is exact and uncompressed, with one feasible assignment per MPO path; compression/bounding remains a downstream solve-design decision.

File: src/projection_mpo.jl (hunk 8 above)
Comment (bernalde) [reply src/projection_mpo.jl:237]: Addressed in `7732e5c`: added `@inferred` tests for typed and default `projection_mpo`/`projection_mpos` calls. The implementation keeps the single `AbstractConstraint` path and relies on `constraint_sites`/`is_feasible` dispatch for constraint-specific behavior.

File: src/projection_mpo.jl (hunk 9 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Addressed in `7732e5c`: expanded the `projection_mpo` docstring to explain the binary Qudit register, diagonal 1/0 semantics, 1-based site indexing, and exact uncompressed path construction.

File: src/projection_mpo.jl (hunk 11 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Addressed in `7732e5c`: added comments explaining feasible-assignment enumeration and how missing sparse coordinates become pass-through MPO tensors. The subtype-specific helper this was attached to has already been collapsed into the `AbstractConstraint` dispatch path.

File: src/projection_mpo.jl (hunk 12 above)
Comment (bernalde) [reply src/projection_mpo.jl:27]: Addressed in `7732e5c`: `SparseTensorEntry` now has a docstring explaining that the dictionary maps register sites to local basis coordinates for a feasible partial assignment, and `docs/src/api.md` includes it in the internal docs.

File: src/projection_mpo.jl [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+      continue
+    end
+
+    inds = mpo_tensor_indices(site, left_link, right_link)
+    nonzeros = Tuple{Vector{Int},T}[]
+
+    for (path, entry) in enumerate(entry_vec)
+      states = get(entry.coordinates, site_position, nothing)
+      states = isnothing(states) ? (1, 2) : (states,)
+      coefficient = site_position == firstindex(site_vec) ? entry.value : one(T)
```
Comment (bernalde) [inline src/projection_mpo.jl]: Nonblocking: `coefficient` is only set to `entry.value` when `site_position == firstindex(site_vec)`. But if the first register site is a pass-through site for every entry (i.e. no entry has it in `coordinates`), the earlier branch builds it with `pass_through_tensor`, which writes `one(T)` and never sees `entry.value` — so the coefficient is silently dropped. Harmless for projection MPOs (entries always carry `value = one(T)`), but `SparseTensorEntry`/`itensor_from_nonzeros` are documented public utilities in api.md, so a caller assembling an MPO with a non-unit value on a pass-through first site would lose it. Either document that `tensor_to_mpo` requires unit values, or thread the coefficient onto the first *constrained* site instead of the first register site.

**Review by bernalde — COMMENTED** [review]
Maintainer review of #85 (native projection-MPO assembly utilities).

I authored this PR, so GitHub only lets me submit this as a `COMMENT`; it is not a formal approval. A prior maintainer review by @iagoleal already set `CHANGES_REQUESTED`, and that gate stands regardless of this comment.

## Verification

- Reviewed the diff against the merge-base with `main`; the changed-file set matches `gh pr diff --name-only` (6 files).
- Ran the targeted tests in a clean worktree at head `7732e5c`: `julia --project=. -e 'using Pkg; Pkg.instantiate(); using Test, TenSolver; include("test/projection_mpo.jl")'` → **45/45 pass** (15.9s). The `@inferred` type-stability cases (which address @iagoleal's typed-vs-untyped question) pass.
- CI on this head is green across all platform jobs and the `build` (docs) job succeeded, so the docs-build failure I flagged on the earlier head is resolved by `7732e5c` ("fix: document projection MPO internals"). All five bindings added to `docs/src/api.md` (`SparseTensorEntry`, `itensor_from_nonzeros`, `projection_mpo`, `projection_mpos`, `constraint_sites`) carry docstrings, so `makedocs` has no missing-docs error.

## Linked issue

`Closes #57` is appropriate: the PR satisfies all four acceptance criteria (diagonal-mask materialization, feasible-preserving/infeasible-zeroing masks, no CoTenN-Python/fork dependency, and the CoTenN attribution comment at the top of `src/projection_mpo.jl`). Manual, per-constraint, and infeasible-mask tests all exercise these.

## Findings

### Nonblocking — stale "Branch Hygiene" section in the PR body 

The PR body still says the base is `pr-1-constraint-api`, that the PR is "stacked on PR #84 … opened as draft", and lists "Prerequisite PRs: #84". Live state contradicts all of this: the base is now `main`, #84 merged on 2026-06-26, and the PR is not a draft. The diff against `main` is clean and CI is green, so the retarget worked — but please update the body so reviewers don't chase a superseded stacking plan.

### Nonblocking — `tensor_to_mpo` silently drops `entry.value` on a pass-through first site

See the inline comment on `src/projection_mpo.jl:124`. This does not affect the shipped feature (projection entries always carry `value = one(T)`), but `SparseTensorEntry`/`itensor_from_nonzeros` are documented in `api.md` as advanced-user utilities, and this is a latent limitation of the generic assembly path.

## Alignment with the existing `CHANGES_REQUESTED`

I concur with @iagoleal's direction: the repeated typed/untyped method pairs and manual tensor-index plumbing (`mpo_tensor_indices`, `mpo_coordinate`, `path_count`, `pass_through_tensor`) are candidates for consolidation, several internal helpers lack docstrings, and the `SparseTensorEntry.value` field is only ever `one(T)` in practice — which is the same "why does this type carry a value?" question raised in review. Worth checking whether ITensors.jl/ITensorMPS.jl provide higher-level constructors before hardening this bespoke path.

## Summary

- Blocking: none from me. The existing `CHANGES_REQUESTED` from @iagoleal is the active gate.
- Nonblocking: stale PR-body branch-hygiene section; `entry.value` drop on a pass-through first site.
- Questions: none.
- Tests: `test/projection_mpo.jl` 45/45 pass locally at head `7732e5c`; full CI green.
- Merge-readiness: code and tests are in good shape and CI is green, but this stays unmergeable until the maintainer `CHANGES_REQUESTED` is resolved. This account cannot provide the formal approval.
